File: CarteComm/A4_Wifi/poste-unipi/agv_poste/esp3.py
"""Décodeur EnOcean Serial Protocol 3 (ESP3) — récepteur TCM 515.

Miroir Python du décodeur C++ du dossier SMS_EnOcean. Ici il tourne sur le
poste fixe UniPi, seul endroit de l'architecture Wi-Fi où de l'EnOcean circule.

Trame ESP3 :
    0x55 | DataLen(2, BE) | OptLen(1) | PacketType(1) | CRC8H |
    Data[DataLen] | OptData[OptLen] | CRC8D

⚠ Le TCM 515 est RÉCEPTION SEULE : aucun retour n'est possible vers le bouton.
L'accusé opérateur passe donc soit par un voyant câblé au poste, soit par un
actionneur EnOcean distinct (planification §3.6 et registre des risques).
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

SYNC = 0x55
TYPE_RADIO_ERP1 = 0x01
RORG_RPS = 0xF6
DATA_MAX = 64
OPT_MAX = 16


def crc8(data: bytes, seed: int = 0x00) -> int:
    """CRC8 EnOcean, polynôme 0x07."""
    crc = seed
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = ((crc << 1) ^ 0x07) & 0xFF if crc & 0x80 else (crc << 1) & 0xFF
    return crc


@dataclass(slots=True)
class Packet:
    packet_type: int
    data: bytes
    opt: bytes
    rssi_dbm: int = 0


@dataclass(slots=True)
class RpsTelegram:
    sender_id: int      # identifiant 32 bits gravé en usine
    data: int
    status: int
    rssi_dbm: int
    pressed: bool       # bit 4 : energy bow armé
    rocker: int         # bits 7..5 : bascule actionnée


class _State(IntEnum):
    SYNC = 0
    HEADER = 1
    CRC_HEADER = 2
    PAYLOAD = 3
    CRC_DATA = 4


class Esp3Decoder:
    """Décodeur incrémental : un octet à la fois, resynchronisation automatique."""

    def __init__(self) -> None:
        self._state = _State.SYNC
        self._header = bytearray()
        self._payload = bytearray()
        self._data_len = 0
        self._opt_len = 0
        self._packet_type = 0
        self.crc_header_errors = 0
        self.crc_data_errors = 0
        self.packets_ok = 0

    def feed(self, byte: int) -> Packet | None:
        if self._state is _State.SYNC:
            if byte == SYNC:
                self._header.clear()
                self._state = _State.HEADER
            return None

        if self._state is _State.HEADER:
            self._header.append(byte)
            if len(self._header) == 4:
                self._state = _State.CRC_HEADER
            return None

        if self._state is _State.CRC_HEADER:
            if crc8(bytes(self._header)) != byte:
                self.crc_header_errors += 1
                self._state = _State.SYNC
                return None
            self._data_len = (self._header[0] << 8) | self._header[1]
            self._opt_len = self._header[2]
            self._packet_type = self._header[3]
            if self._data_len > DATA_MAX or self._opt_len > OPT_MAX:
                self._state = _State.SYNC
                return None
            self._payload.clear()
            self._state = _State.PAYLOAD if (self._data_len + self._opt_len) else _State.CRC_DATA
            return None

        if self._state is _State.PAYLOAD:
            self._payload.append(byte)
            if len(self._payload) == self._data_len + self._opt_len:
                self._state = _State.CRC_DATA
            return None

        # _State.CRC_DATA
        self._state = _State.SYNC
        if crc8(bytes(self._payload)) != byte:
            self.crc_data_errors += 1
            return None

        data = bytes(self._payload[: self._data_len])
        opt = bytes(self._payload[self._data_len:])
        # OptData d'un RADIO_ERP1 : [subTelNum][destinationID×4][dBm][security]
        rssi = -opt[5] if self._packet_type == TYPE_RADIO_ERP1 and len(opt) >= 6 else 0
        self.packets_ok += 1
        return Packet(self._packet_type, data, opt, rssi)


def parse_rps(packet: Packet) -> RpsTelegram | None:
    """Extrait un télégramme RPS (PTM 210). None si ce n'en est pas un."""
    if packet.packet_type != TYPE_RADIO_ERP1:
        return None
    if len(packet.data) < 7 or packet.data[0] != RORG_RPS:
        return None
    return RpsTelegram(
        sender_id=int.from_bytes(packet.data[2:6], "big"),
        data=packet.data[1],
        status=packet.data[6],
        rssi_dbm=packet.rssi_dbm,
        pressed=bool(packet.data[1] & 0x10),
        rocker=(packet.data[1] >> 5) & 0x03,
    )


class Deduplicator:
    """Le PTM 210 émet 3 sous-télégrammes identiques par appui.

    Sans ce filtre, un appui déclencherait trois courses.
    """

    def __init__(self, window_ms: int = 100, slots: int = 8) -> None:
        self._window_ms = window_ms
        self._slots: list[tuple[int, int, float]] = []
        self._max = slots
        self.duplicates = 0

    def accept(self, sender_id: int, data: int, now_ms: float) -> bool:
        for slot_id, slot_data, at in self._slots:
            if slot_id == sender_id and slot_data == data and (now_ms - at) <= self._window_ms:
                self.duplicates += 1
                return False
        self._slots.append((sender_id, data, now_ms))
        if len(self._slots) > self._max:
            self._slots.pop(0)
        return True

File: Comm distance/architectures/A2_Hybride/firmware/common/enocean/esp3.cpp
#include "enocean/esp3.h"

#include "proto/crc16.h"

namespace agv {

bool Esp3Decoder::feed(uint8_t byte, Esp3Packet& out) {
  switch (state_) {
    case State::Sync:
      if (byte == kEsp3Sync) {
        index_ = 0;
        state_ = State::Header;
      }
      break;

    case State::Header:
      header_[index_++] = byte;
      if (index_ == 4) state_ = State::CrcHeader;
      break;

    case State::CrcHeader: {
      if (crc8_enocean(header_, 4) != byte) {
        ++crc_header_errors_;
        state_ = State::Sync;  // resynchronisation sur le prochain 0x55
        break;
      }
      current_ = Esp3Packet{};
      current_.data_len = static_cast<uint16_t>((header_[0] << 8) | header_[1]);
      current_.opt_len = header_[2];
      current_.packet_type = header_[3];
      if (current_.data_len > kEsp3DataMax || current_.opt_len > kEsp3OptMax) {
        state_ = State::Sync;  // trame plus longue que ce qu'on sait traiter
        break;
      }
      payload_expected_ = current_.data_len + current_.opt_len;
      index_ = 0;
      state_ = (payload_expected_ == 0) ? State::CrcData : State::Payload;
      break;
    }

    case State::Payload:
      if (index_ < current_.data_len) {
        current_.data[index_] = byte;
      } else {
        current_.opt[index_ - current_.data_len] = byte;
      }
      ++index_;
      if (index_ == payload_expected_) state_ = State::CrcData;
      break;

    case State::CrcData: {
      uint8_t buf[kEsp3DataMax + kEsp3OptMax];
      size_t n = 0;
      for (size_t i = 0; i < current_.data_len; ++i) buf[n++] = current_.data[i];
      for (size_t i = 0; i < current_.opt_len; ++i) buf[n++] = current_.opt[i];
      state_ = State::Sync;
      if (crc8_enocean(buf, n) != byte) {
        ++crc_data_errors_;
        return false;
      }
      // OptData d'un RADIO_ERP1 : [subTelNum][destinationID×4][dBm][security]
      if (current_.packet_type == kEsp3TypeRadioErp1 && current_.opt_len >= 6) {
        current_.rssi_dbm = -static_cast<int8_t>(current_.opt[5]);
      }
      ++packets_ok_;
      out = current_;
      return true;
    }
  }
  return false;
}

bool Esp3Decoder::parse_rps(const Esp3Packet& p, RpsTelegram& out) {
  // RPS : RORG(1) + data(1) + senderID(4) + status(1) = 7 octets.
  if (p.packet_type != kEsp3TypeRadioErp1) return false;
  if (p.data_len < 7 || p.data[0] != kRorgRps) return false;

  out.data = p.data[1];
  out.sender_id = (static_cast<uint32_t>(p.data[2]) << 24) |
                  (static_cast<uint32_t>(p.data[3]) << 16) |
                  (static_cast<uint32_t>(p.data[4]) << 8) | static_cast<uint32_t>(p.data[5]);
  out.status = p.data[6];
  out.rssi_dbm = p.rssi_dbm;
  // Bit 4 (energy bow) : 1 = appui, 0 = relâchement.
  out.pressed = (out.data & 0x10u) != 0u;
  // Bits 7..5 : identifiant du bascule actionné (R1).
  out.rocker = static_cast<uint8_t>((out.data >> 5) & 0x03u);
  return true;
}

bool EnoceanDeduplicator::accept(uint32_t sender_id, uint8_t data, uint32_t now_ms) {
  for (const auto& slot : slots_) {
    if (slot.used && slot.sender_id == sender_id && slot.data == data &&
        (now_ms - slot.at_ms) <= window_ms_) {
      ++duplicates_;
      return false;
    }
  }
  slots_[next_] = Slot{sender_id, data, now_ms, true};
  next_ = (next_ + 1) % kSlots;
  return true;
}

}  // namespace agv
